server/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java

```diff
@@ -692,10 +694,16 @@ private static void parseNonDynamicArray(
                     || objectMapper.dynamic == ObjectMapper.Dynamic.RUNTIME);
             boolean fieldWithFallbackSyntheticSource = mapper instanceof FieldMapper fieldMapper
                 && fieldMapper.syntheticSourceMode() == FieldMapper.SyntheticSourceMode.FALLBACK;
-            boolean fieldWithStoredArraySource = mapper instanceof FieldMapper fieldMapper
+            boolean fieldWithStoredArraySource = mapper instanceof FieldMapper
                 && context.sourceKeepModeFromIndexSettings() == Mapper.SourceKeepMode.ARRAYS;
             boolean dynamicRuntimeContext = context.dynamic() == ObjectMapper.Dynamic.RUNTIME;
-            if (objectRequiresStoringSource || fieldWithFallbackSyntheticSource || dynamicRuntimeContext || fieldWithStoredArraySource) {
+            boolean copyToFieldHasValuesInDocument = context.isWithinCopyTo() == false
```

Comment:
Nit: `String fullPath = context.path().pathAsText(arrayFieldName);' and use below.